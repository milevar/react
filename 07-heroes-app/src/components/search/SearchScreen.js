import {useForm} from "../../hooks/useForm";
import {getHeroesByName} from "../../selectors/getHeroesByName";
import {HeroCard} from "../hero/HeroCard";
import {useLocation, useNavigate} from "react-router-dom";
import queryString from "query-string";
import {useMemo} from "react";

export const SearchScreen = () => {

    const navigate = useNavigate();
    const location = useLocation();

    const {q = ''} = queryString.parse(location.search);

    const [formValues, handleInputChange] = useForm({
        searchText: q,
    });

    const {searchText} = formValues;
    const heroesFiltered = useMemo(() => getHeroesByName(q), [q]);;

    const handleSearch = (e) => {
        e.preventDefault();
        console.log(searchText)
        navigate(`?q=${searchText}`);

    }

    return (
        <>
            <h1>Search a Hero</h1>
            <hr/>

            <div className="row">
                <div className="col-5">
                    <h4>Search</h4>
                    <hr/>
                    <form onSubmit={handleSearch}>
                        <input
                            type="text"
                            placeholder="Search a hero"
                            className="form-control"
                            name="searchText"
                            autoComplete="off"
                            value={searchText}
                            onChange={handleInputChange}
                        />

                        <button className="btn btn-outline-primary mt-1"
                                type="submit">
                            Search</button>
                    </form>
                </div>
                <div className="col-5">
                    <h4>Results</h4>
                    <hr/>
                    {
                        (q === '')
                        ? <div className="alert alert-info">You should to search a hero.</div>
                        : (heroesFiltered.length === 0)
                            && <div className="alert alert-danger">No found results to: {q} </div>
                    }

                    {
                        heroesFiltered.map(hero => (
                            <HeroCard
                                key={hero.id}
                                { ...hero}
                            />
                        ))
                    }
                </div>
            </div>


        </>
    );

}
