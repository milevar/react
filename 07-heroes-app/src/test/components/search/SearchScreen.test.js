import {mount} from "enzyme";
import {SearchScreen} from "../../../components/search/SearchScreen";
import {MemoryRouter, useNavigate} from "react-router-dom";

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
        useNavigate: () => mockNavigate,
}));

describe('Pruebas en <SearchScreen />', () => {


    test('debe de mostrarse correctamente con valores por defecto', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}  >
                <SearchScreen />
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').text().trim()).toBe('You should to search a hero.');
    })

    test('debe de mostrar a Batman y el input con el valor de queryString', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}  >
                <SearchScreen />
            </MemoryRouter>
        );

        expect(wrapper.find('input').prop('value')).toBe('batman');
        expect(wrapper).toMatchSnapshot();
    })

    test('debe de mostrar un error si no se encuentra el hero', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman123']}  >
                <SearchScreen />
            </MemoryRouter>
        );

        expect(wrapper.find('.alert-danger').text().trim()).toBe('No found results to: batman123');

    })

    test('debe de llamar el navite a la nueva pantalla', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}  >
                <SearchScreen />
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change', {
            target:{
                name:'searchText',
                value:'batman'
            }
        });

        wrapper.find('form').prop('onSubmit')({
            preventDefault: () => {}
        })

        expect(mockNavigate).toHaveBeenCalledWith('?q=batman');

    })

})
