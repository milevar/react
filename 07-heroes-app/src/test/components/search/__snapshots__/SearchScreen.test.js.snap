// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen /> debe de mostrar a Batman y el input con el valor de queryString 1`] = `
Array [
  <h1>
    Search a Hero
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search a hero"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-outline-primary mt-1"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-5"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="col animate__animated animate__fadeInLeft"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Batman"
                className="card-img-top"
                src="/assets/heroes/dc-batman.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Batman
                </h5>
                <p
                  className="card-text"
                >
                  Bruce Wayne
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  >
                    Detective Comics #27
                  </small>
                </p>
                <a
                  href="/hero/dc-batman"
                  onClick={[Function]}
                >
                  Mas ...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`Pruebas en <SearchScreen /> debe de mostrarse correctamente con valores por defecto 1`] = `
Array [
  <h1>
    Search a Hero
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary mt-1"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-5"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        You should to search a hero.
      </div>
    </div>
  </div>,
]
`;
